File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_cpp/code_output/BinaryDataProcessor.cpp
#include <iostream>
#include <string>
#include <regex>
#include <algorithm>
#include <sstream>
#include <iomanip>
#include <stdexcept>

class BinaryInfo {
public:
    double zeroes;
    double ones;
    int bitLength;

    BinaryInfo(double zeroes, double ones, int bitLength)
        : zeroes(zeroes), ones(ones), bitLength(bitLength) {}

    std::string toString() const {
        std::ostringstream oss;
        oss << std::fixed << std::setprecision(3);
        oss << "{Zeroes: " << zeroes << ", Ones: " << ones << ", Bit length: " << bitLength << "}";
        return oss.str();
    }
};

std::ostream& operator<<(std::ostream& os, const BinaryInfo& info) {
    os << info.toString();
    return os;
}

class BinaryDataProcessor {
private:
    std::string binaryString;

    void cleanNonBinaryChars() {
        binaryString = std::regex_replace(binaryString, std::regex("[^01]"), "");
    }

public:
    BinaryDataProcessor(const std::string& input) : binaryString(input) {
        cleanNonBinaryChars();
    }

    BinaryInfo calculateBinaryInfo() const {
        int zeroesCount = std::count(binaryString.begin(), binaryString.end(), '0');
        int onesCount = std::count(binaryString.begin(), binaryString.end(), '1');
        int totalLength = binaryString.length();

        double zeroesPercentage = static_cast<double>(zeroesCount) / totalLength;
        double onesPercentage = static_cast<double>(onesCount) / totalLength;

        return BinaryInfo(zeroesPercentage, onesPercentage, totalLength);
    }

    std::string convertToAscii() const {
        std::string asciiString;
        for (size_t i = 0; i < binaryString.length(); i += 8) {
            if (i + 8 > binaryString.length()) {
                throw std::out_of_range("Binary string length is not a multiple of 8");
            }
            std::string byteString = binaryString.substr(i, 8);
            int decimal = std::stoi(byteString, nullptr, 2);
            asciiString += static_cast<char>(decimal);
        }
        return asciiString;
    }

    std::string convertToUtf8() const {
        // Identical to convertToAscii in this implementation
        std::string utf8String;
        for (size_t i = 0; i < binaryString.length(); i += 8) {
            if (i + 8 > binaryString.length()) {
                throw std::out_of_range("Binary string length is not a multiple of 8");
            }
            std::string byteString = binaryString.substr(i, 8);
            int decimal = std::stoi(byteString, nullptr, 2);
            utf8String += static_cast<char>(decimal);
        }
        return utf8String;
    }

    const std::string& getBinaryString() const {
        return binaryString;
    }
};

int main() {
    BinaryDataProcessor bdp("0110100001100101011011000110110001101111");
    std::cout << bdp.getBinaryString() << std::endl;
    std::cout << bdp.calculateBinaryInfo() << std::endl;
    std::cout << bdp.convertToAscii() << std::endl;
    std::cout << bdp.convertToUtf8() << std::endl;
    return 0;
}

File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/java_to_py/BinaryDataProcessor.py
class BinaryDataProcessor:
    def __init__(self, binary_string):
        self.binary_string = binary_string
        self.clean_non_binary_chars()

    def clean_non_binary_chars(self):
        self.binary_string = ''.join(bit for bit in self.binary_string if bit in '01')

    def calculate_binary_info(self):
        total_length = len(self.binary_string)
        if total_length == 0:
            return BinaryInfo(0.0, 0.0, 0)
        
        zeros_count = total_length - len(self.binary_string.replace('0', ''))
        ones_count = total_length - len(self.binary_string.replace('1', ''))
        
        zeroes_percentage = zeros_count / total_length
        ones_percentage = ones_count / total_length
        
        return BinaryInfo(zeroes_percentage, ones_percentage, total_length)

    def convert_to_ascii(self):
        result = []
        for i in range(0, len(self.binary_string), 8):
            byte_str = self.binary_string[i:i+8]
            if byte_str == '':
                continue
            decimal = int(byte_str, 2)
            result.append(chr(decimal))
        return ''.join(result)

    def get_binary_string(self):
        return self.binary_string


class BinaryInfo:
    def __init__(self, zeroes, ones, bit_length):
        self.zeroes = zeroes
        self.ones = ones
        self.bit_length = bit_length

    def __str__(self):
        return f"{{Zeroes: {self.zeroes:.3f}, Ones: {self.ones:.3f}, Bit length: {self.bit_length}}}"


if __name__ == "__main__":
    bdp = BinaryDataProcessor("0110100001100101011011000110110001101111")
    print(bdp.get_binary_string())
    print(bdp.calculate_binary_info())
    print(bdp.convert_to_ascii())